File: pkg/geometry/disk.go
package geometry

import (
	"math"

	"sif/gogs/eric/relray/pkg/vec"
)

// Disk is a flat circular surface on the XZ plane (Y=0) with normal pointing +Y.
// Use a Transformed wrapper to position and orient it in the scene.
type Disk struct {
	Radius float64
}

func (d *Disk) Intersect(origin, dir vec.Vec3, tMin, tMax float64) (Hit, bool) {
	if math.Abs(dir.Y) < 1e-12 {
		return Hit{}, false
	}
	t := -origin.Y / dir.Y
	if t < tMin || t > tMax {
		return Hit{}, false
	}
	p := origin.Add(dir.Scale(t))
	if p.X*p.X+p.Z*p.Z > d.Radius*d.Radius {
		return Hit{}, false
	}
	h := Hit{T: t, Point: p}
	h.SetFaceNormal(dir, vec.Vec3{Y: 1})
	return h, true
}

File: rust/src/geometry/disk.rs
use crate::vec::Vec3;
use super::{Hit, Shape};

pub struct Disk {
    pub radius: f64,
}

impl Shape for Disk {
    fn intersect(&self, origin: Vec3, dir: Vec3, t_min: f64, t_max: f64) -> Option<Hit> {
        if dir.y.abs() < 1e-12 {
            return None;
        }
        let t = -origin.y / dir.y;
        if t < t_min || t > t_max {
            return None;
        }
        let p = origin + dir * t;
        if p.x * p.x + p.z * p.z > self.radius * self.radius {
            return None;
        }
        let mut h = Hit { t, point: p, ..Default::default() };
        h.set_face_normal(dir, Vec3 { x: 0.0, y: 1.0, z: 0.0 });
        Some(h)
    }
}
